src/base/http/connection.cpp

```diff
@@ -180,9 +180,96 @@ QHostAddress Connection::resolvePeerAddress(const Http::Request &request)
     if (reverseProxyAddress.setAddress(reverseProxyAddressString) && peerAddress.isEqual(reverseProxyAddress)) {
         QString forwardedFor = request.headers.value(Http::HEADER_X_FORWARDED_FOR);
 
-        if (!forwardedFor.isEmpty())
-            peerAddress.setAddress(forwardedFor.split(",").at(0));
+        if (!forwardedFor.isEmpty()) {
+            // peer address is the 1st global IP in X-Forwarded-For or, if none available, the 1st IP in the list
+            QStringList remoteIpList= forwardedFor.split(",");
+            bool hasGlobalIp = false;
+
+            foreach (const QString &remoteIp, remoteIpList) {
+               if (peerAddress.setAddress(remoteIp) && isGlobal(peerAddress)) {
+                   hasGlobalIp = true;
+                   break;
+               }
+            }
+
+            if (!hasGlobalIp)
+                peerAddress.setAddress(remoteIpList.at(0));
+        }
     }
 
     return peerAddress;
 }
+
+bool Connection::isGlobal(const QHostAddress ip)
+{
+#if QT_VERSION < QT_VERSION_CHECK(5, 11, 0)
+    // check IP against IPv4 special addresses
+    if (ip.isInSubnet(QHostAddress("10.0.0.0"), 8))
+        return false;
+
+    if (ip.isInSubnet(QHostAddress("100.64.0.0"), 10))
+        return false;
+
+    if (ip.isInSubnet(QHostAddress("127.0.0.0"), 8))
+        return false;
+
+    if (ip.isInSubnet(QHostAddress("169.254.0.0"), 16))
+        return false;
+
+    if (ip.isInSubnet(QHostAddress("172.16.0.0"), 12))
+        return false;
+
+    if (ip.isInSubnet(QHostAddress("192.0.0.0"), 24))
```

Comment:
What's the significance of most of these IPs? For example, `192.0.0.0/24` is a mixture of private and public IPs.

Reply:
> For example, 192.0.0.0/24 is a mixture of private and public IPs.

??? which part is private and which is public between 192.0.0.0 and 192.0.0.255 ?

You can find the list of these IP ranges on wikipedia, just search "reserved ip address". It contains the well known 10.x, 169.254.x and 192.168.x but also several special ip ranges which either are reserved for future use or without meaning for bittorrent (multicast range).

I whould have prefered to use QT5 function isGlobal but it requires a very recent QT (5.11) not yet available on most distros (mine included)